server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -561,19 +564,30 @@ private void balance() {
         }
 
         private void maybeLogUndesiredAllocationsWarning(int allAllocations, int undesiredAllocations, int nodeCount) {
-            // more shards than cluster can relocate with one reroute
-            final boolean nonEmptyRelocationBacklog = undesiredAllocations > 2L * nodeCount;
-            final boolean warningThresholdReached = undesiredAllocations > undesiredAllocationsLogThreshold * allAllocations;
-            if (allAllocations > 0 && nonEmptyRelocationBacklog && warningThresholdReached) {
-                undesiredAllocationLogInterval.maybeExecute(
-                    () -> logger.warn(
-                        "[{}] of assigned shards ({}/{}) are not on their desired nodes, which exceeds the warn threshold of [{}]",
+            // There are more shards in undesirable locations than the cluster can relocate with one reroute per shard.
```

Comment:
I am not sure what `one reroute per shard` menas

Reply:
Hmm. I think I meant to say "node" at the end, not "shard".

On further thought, I really don't understand what the original comment meant. I see that you originally wrote the comment: could you elaborate on what nonEmptyRelocationBacklog (in the original code) is intended to mean?

Reply:
Please see the pr level comment clarifying this point

Reply:
So, IIUC, `nonEmptyRelocationBacklog` was an attempt to log the warning ONLY if the balancer has decided to move a non trivial number of shards? `2L * nodeCount` is saying that the balancer wants to move more than `1L * nodeCount`. Presumably in connection with some balancer algorithm detail. Or perhaps simply avoiding logging anything if the the old balancer algorithm is active that would only plan one step at a time?